File: benchmarks/python/05_range_loop.py
import sqlite3
import time
import os
import sys
from bench_utils import get_connection, load_plsqlite

def bench_python_range(n):
    conn = get_connection()
    conn.execute("DROP TABLE IF EXISTS range_test")
    conn.execute("CREATE TABLE range_test(id INTEGER PRIMARY KEY, val INTEGER)")
    start = time.time()
    conn.execute("BEGIN TRANSACTION")
    for i in range(1, n + 1):
        conn.execute("INSERT INTO range_test(val) VALUES (?)", (i,))
    conn.commit()
    end = time.time()
    res = conn.execute("SELECT count(*) FROM range_test").fetchone()[0]
    conn.close()
    return (end - start) * 1000, res

def bench_plsqlite_range(n):
    conn = get_connection()
    load_plsqlite(conn)
    conn.execute("DROP TABLE IF EXISTS range_test")
    conn.execute("CREATE TABLE range_test(id INTEGER PRIMARY KEY, val INTEGER)")

    conn.execute(f"""
    SELECT register_plsql('range_insert_test', 'n', '
        RANGE i IN (1, @n) LOOP
            INSERT INTO range_test(val) VALUES (@i);
        END LOOP;
    ');
    """)

    start = time.time()
    conn.execute("SELECT run_plsql('range_insert_test', ?);", (n,)).fetchone()
    end = time.time()
    res = conn.execute("SELECT count(*) FROM range_test").fetchone()[0]
    conn.close()
    return (end - start) * 1000, res

if __name__ == "__main__":
    N = 10000 # Use 10k for insert benchmark
    print(f"--- RANGE-Insert Benchmarks ({N:,} iterations) ---")

    py_time, py_res = bench_python_range(N)
    print(f"RESULT: Range: App-Layer: {py_time:.2f}ms (Result: {py_res})")

    pl_time, pl_res = bench_plsqlite_range(N)
    print(f"RESULT: Range: PL/SQLite: {pl_time:.2f}ms (Result: {pl_res})")

    if pl_time > 0:
        print(f"Speedup: {py_time/pl_time:.2f}x")
    else:
        print("Speedup: N/A (PL/SQLite time was 0ms)")

File: benchmarks/node/05_range_loop.js
const { getDbConnection, loadPLSQLite } = require('./bench_utils');

function benchNodeRange(n) {
    const db = getDbConnection();
    db.prepare("DROP TABLE IF EXISTS range_test").run();
    db.prepare("CREATE TABLE range_test(id INTEGER PRIMARY KEY, val INTEGER)").run();
    const start = Date.now();
    db.transaction((n) => {
        const stmt = db.prepare("INSERT INTO range_test(val) VALUES (?)");
        for (let i = 1; i <= n; i++) {
            stmt.run(i);
        }
    })(n);
    const end = Date.now();
    const count = db.prepare("SELECT count(*) as c FROM range_test").get().c;
    db.close();
    return { time: end - start, total: count };
}

function benchPLSQLiteRange(n) {
    const db = getDbConnection();
    loadPLSQLite(db);
    db.prepare("DROP TABLE IF EXISTS range_test").run();
    db.prepare("CREATE TABLE range_test(id INTEGER PRIMARY KEY, val INTEGER)").run();

    db.prepare(`
        SELECT register_plsql('range_insert_test', 'n', '
            RANGE i IN (1, @n) LOOP
                INSERT INTO range_test(val) VALUES (@i);
            END LOOP;
        ');
    `).run();

    const start = Date.now();
    db.prepare("SELECT run_plsql('range_insert_test', ?)").run(n);
    const end = Date.now();
    const count = db.prepare("SELECT count(*) as c FROM range_test").get().c;
    db.close();
    return { time: end - start, total: count };
}

const N = 10000;
console.log(`--- RANGE-Insert Benchmarks (${N.toLocaleString()} iterations) ---`);

const nodeRes = benchNodeRange(N);
console.log(`RESULT: Range: App-Layer: ${nodeRes.time.toFixed(2)}ms (Result: ${nodeRes.total})`);

const plRes = benchPLSQLiteRange(N);
console.log(`RESULT: Range: PL/SQLite: ${plRes.time.toFixed(2)}ms (Result: ${plRes.total})`);

if (plRes.time > 0) {
    console.log(`Speedup: ${(nodeRes.time / plRes.time).toFixed(2)}x`);
}
